src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -678,7 +678,7 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             var errorCode = (Http2ProtocolErrorCode)BinaryPrimitives.ReadInt32BigEndian(_incomingBuffer.ActiveSpan.Slice(sizeof(int)));
             if (NetEventSource.IsEnabled) Trace(frameHeader.StreamId, $"{nameof(lastValidStream)}={lastValidStream}, {nameof(errorCode)}={errorCode}");
 
-            AbortStreams(lastValidStream, new Http2ProtocolException(errorCode));
+            StartTerminatingConnection(lastValidStream, new Http2ProtocolException(errorCode));
```

Comment:
I'm not sure it's worthwhile passing the Http2ProtocolException through here. The requests are going to be retried, and thus the user should never see this exception. I suppose it doesn't cause any harm, but it is different than how we handle REFUSED_STREAM today.

Reply:
@geoffkizer the reason I added that was so that user have a way of distinguishing between NO_ERROR vs other state in case they implement their own retry logic. Also I've noticed the initially written test expected to see that being propagated so it's possible the user might want to have more diagnostic info as well. (also even though retry logic is there after certain amount of tries/time we still bubble up the exception which looks more useful with extra info)

Reply:
if the user is seeing this exception, then there's something wrong going on. We should always retry when this exception is thrown. Do you know why this is getting surfaced to the user?

This may be related to the issue below where you are sending a GOAWAY with last valid stream = 0. I.e., the server is lying and telling us that requests which have already been processed actually have not been, and since we've gotten past the point in request processing where we can retry (i.e. response headers have been received) then we're actually ending up aborting this request. I guess that's fine, but at least we should add a comment regarding this...

Reply:
Since you are passing this exception in the GOAWAY case, can we also pass it in the REFUSED_STREAM case?  Seems like we could have a similar issue with the server incorrectly sending REFUSED_STREAM after it's processed the stream.

Reply:
I've noticed this being bubbled when timeout happens which I actually think it makes sense although I do not think the code was designed to do that and likely is a coincidence (i.e. client tried doing request, retried few times but server kept on aborting with some error and eventually you see one of the server sent bubbled up).

Reply:
That's strange. Do you know which test this was happening during? This was a remote server test, I assume? It seems weird that we'd be getting REFUSED_STREAM at all during remote server tests... would be good to understand why.

Reply:
I'll pass it also in the REFUSED_STREAM case. Unfortunately wasn't able to repro bubbling up condition since I got it last time so possibly this was due to some bug I got when I was modifying the code.